rmw_fastrtps_shared_cpp/src/rmw_service_server_is_available.cpp

```diff
@@ -100,11 +100,22 @@ __rmw_service_server_is_available(
     return RMW_RET_OK;
   }
 
-  if (0 == client_info->request_publisher_matched_count_.load()) {
+  if (number_of_request_subscribers != number_of_response_publishers) {
```

Comment:
@MiguelCompany why do still check for the global count? Isn't checking for the matched count enough? Won't this make this check fail for a fully matched client if another client for the same service is still halfway through discovery?

Reply:
I assume we want to ensure that the discovered servers are consistent. This check will fail if there is a **server** halfway through discovery.

I guess it is ok if we just check the matched count for equality. It will also be inexact if the client is halfway matched with two servers, but it is an improvement nonetheless.

Reply:
> It will also be inexact if the client is halfway matched with two servers, but it is an improvement nonetheless.

That is a good point, but AFAIK the semantics of multiple servers for the same service in ROS 2 aren't well defined. I wouldn't worry as much about that use case. @wjwwood @ivanpauno for valuable input.

Reply:
Yes, I wouldn't worry about multiple servers behavior now.
I think that's another point where we can iterate in a future release.

Reply:
Shall I revert 2590645a9de94ad5582415f9c7f258505ef9dc74 then?

Reply:
No, that is correct. What seemed redundant were the graph cache checks that were already there. Fine to leave as-is, failure conditions are much more unlikely now and will ultimately get caught by the check in `rmw_send_response`.